core/CronArchive.php

```diff
@@ -1383,169 +829,176 @@ private function getApiToInvalidateArchivedReport()
 
     public function invalidateArchivedReportsForSitesThatNeedToBeArchivedAgain()
     {
+        $this->logger->info("Checking for queued invalidations...");
+
+        // invalidate remembered site/day pairs
         $sitesPerDays = $this->invalidator->getRememberedArchivedReportsThatShouldBeInvalidated();
 
         foreach ($sitesPerDays as $date => $siteIds) {
             //Concurrent transaction logic will end up with duplicates set.  Adding array_unique to the siteIds.
-            $listSiteIds = implode(',', array_unique($siteIds ));
+            $listSiteIds = implode(',', array_unique($siteIds));
 
             try {
-                $this->logger->info('- Will invalidate archived reports for ' . $date . ' for following websites ids: ' . $listSiteIds);
+                $this->logger->info('  Will invalidate archived reports for ' . $date . ' for following websites ids: ' . $listSiteIds);
                 $this->getApiToInvalidateArchivedReport()->invalidateArchivedReports($siteIds, $date);
             } catch (Exception $e) {
-                $this->logger->info('Failed to invalidate archived reports: ' . $e->getMessage());
+                $this->logger->info('  Failed to invalidate archived reports: ' . $e->getMessage());
             }
         }
-    }
 
-    /**
-     *  Returns the list of sites to loop over and archive.
-     *  @return array
-     */
-    public function initWebsiteIds()
-    {
-        if (count($this->shouldArchiveSpecifiedSites) > 0) {
-            $this->logger->info("- Will process " . count($this->shouldArchiveSpecifiedSites) . " websites (--force-idsites)");
+        // invalidate today if needed for all websites
+        foreach ($this->allWebsites as $idSite) {
+            $timezone = Site::getTimezoneFor($idSite);
+            $today = Date::factoryInTimezone('today', $timezone);
 
-            return $this->shouldArchiveSpecifiedSites;
-        }
+            $todayPeriod = PeriodFactory::build('day', $today);
+
+            $params = new Parameters(new Site($idSite), $todayPeriod, new Segment('', [$idSite]));
+            if ($this->isThereExistingValidPeriod($params)) {
+                $this->logger->debug("  Found existing valid archive for today, skipping invalidation...");
+                continue;
+            }
+
+            $loader = new Loader($params);
+            if ($loader->canSkipThisArchive()) {
+                $this->logger->debug("  Today archive can be skipped due to no visits, skipping invalidation...");
+                continue;
+            }
```

Comment:
Explanation part 7: above are two optimizations done for some archives. isThereExistingValidPeriod() checks if there's a usable archive using a ttl that depends on whether the period includes today.

canSkipThisArchive() is a new optimization used in several places. If there are no visits for a period AND no child archives that can be used to aggregate, then we don't have to launch archiving, so we don't need to invalidate.